runtime/lua/vim/lsp/document_color.lua

```diff
@@ -0,0 +1,379 @@
+    or not result
+    or not api.nvim_buf_is_loaded(bufnr)
+    or not bufstate.enabled
+  then
+    return
+  end
+
+  if not bufstate.hl_infos or not bufstate.buf_version then
+    bufstate.hl_infos = {}
+    bufstate.buf_version = ctx.version
+  end
+
+  local hl_infos = {}
+  for _, res in ipairs(result) do
+    local range = {
+      res.range.start.line,
+      res.range.start.character,
+      res.range['end'].line,
+      res.range['end'].character,
+    }
+    local hex_code = get_hex_code(res.color)
+    local hl_info = { range = range, hex_code = hex_code }
+
+    if type(M._opts.style) == 'string' then
+      hl_info.hl_group = get_hl_group(hex_code, M._opts.style --[[@as string]])
+    end
+
+    table.insert(hl_infos, hl_info)
+  end
+  bufstate.hl_infos[client_id] = hl_infos
+
+  bufstate.buf_version = ctx.version
+  api.nvim__redraw({ buf = bufnr, valid = true, flush = false })
+end
+
+--- Query whether document colors are enabled in the filtered scope.
+---
+--- @param filter? vim.lsp.document_color.enable.Filter
+--- @return boolean
+function M.is_enabled(filter)
+  vim.validate('filter', filter, 'table', true)
+
+  filter = filter or {}
+  local bufnr = filter.bufnr
+
+  if bufnr == nil then
+    return global_state.enabled
+  else
+    return bufstates[vim._resolve_bufnr(bufnr)].enabled
+  end
+end
+
+--- Optional filters |kwargs|, or `nil` for all.
+--- @inlinedoc
+--- @class vim.lsp.document_color.enable.Filter
+---
+--- Buffer number, or 0 for current buffer, or nil for all.
+--- @field bufnr? integer
+
+--- Enables document highlighting from the given language client in the given buffer.
+---
+--- You can enable document highlighting from a supporting client as follows:
+--- ```lua
+--- vim.api.nvim_create_autocmd('LspAttach', {
+---   callback = function(args)
+---     local client = vim.lsp.get_client_by_id(args.data.client_id)
+---
+---     if client:supports_method('textDocument/documentColor')
+---       vim.lsp.document_color.enable(true, { bufnr = args.buf })
+---     end
+---   end
+--- })
+--- ```
+---
+--- To "toggle", pass the inverse of `is_enabled()`:
+---
+--- ```lua
+--- vim.lsp.document_color.enable(not vim.lsp.document_color.is_enabled())
+--- ```
+---
+--- @param enable? boolean True to enable, false to disable. Defaults to true.
+--- @param filter? vim.lsp.document_color.enable.Filter
+--- @param opts? vim.lsp.document_color.enable.Opts
+function M.enable(enable, filter, opts)
+  vim.validate('enable', enable, 'boolean', true)
+  vim.validate('filter', filter, 'table', true)
+  vim.validate('opts', opts, 'table', true)
+
+  enable = enable == nil or enable
+  filter = filter or {}
+  M._opts = vim.tbl_extend('keep', opts or {}, M._opts)
+
+  if filter.bufnr == nil then
+    global_state.enabled = enable
+    for _, bufnr in ipairs(api.nvim_list_bufs()) do
+      if api.nvim_buf_is_loaded(bufnr) then
+        if enable then
+          buf_enable(bufnr)
+        else
+          buf_disable(bufnr)
```

Comment:
(out of scope) sure seems like we have boilerplate for enable() functionality that could live in lsp._util